### DISCUSSION TITLE
Reactivity not working in Node.js

### DISCUSSION THREAD

**Discussion by msteen** [solidjs/solid#1124] [Q&A] [answered]
I am using Solid on the front-end, but I also plan to use its reactivity primitives to keep track of changes made in models used within the application. Those models will run in both a browser and Node.js context and after building my first app in Solid I tried using the reactivity primitives standalone within Node.js, but it seems like the tracking context does not work in any way.

I found various examples that should work, and do work within the browser, but fail to trigger effects as expected when run directly in Node.js. For example, in discussion #397 there is a link to [a playground](https://playground.solidjs.com/?version=1.4.1#NobwRAdghgtgpmAXGGUCWEwBowBcCeADgsrgM4Ae2YZA9gK4BOAxiWGjIbY7gAQi9mjOFFxwAogDNJcZriyDhouAGU0Ac2gAbXgF9ekxrRi8AOjVpa0AEwC0AKzLmA3KYjNaEMn2Af6EeV4yOFwAYQYAgF1eAF5FETE1TSgtAAoABgBKVwg3IQSJaVlcVNTM2IA+fjdeQU86LTgAOi1adVS-ALLs3gB6Xt4Ad08Acj5cRg11OEZeAH4Ft11stzdgsIiSgDYVrxDw-xKAdl31g66ADl2wXUigA) that works exactly as I hoped, but when I try to run the exact same in Node.js, it does not behave the same (no effects are triggered).

Is this known behavior and Solid's reactive primitives simply cannot be used outside the browser context, or is this usage never considered?

My plan B is to use `@vue/reactivity`, but it is mostly undocumented and it would force me to learn their internals in more details than I would like.

**Comment by edemaine** [comment, marked as answer]
Yes, this behavior is normal and is necessary for SSR to work. One workaround is to declare the Node condition of `"browser"`, e.g. via `-C browser`. Another is to `import` from `"solid-js/dist/solid.js"` to force getting the browser build (but this doesn't work with libraries that directly import from `"solid-js"`).

#1104 has some proposals for how to make this easier in the future, but we're not there yet.

**Reply by msteen** [reply]
Thanks, that explains it! Guess it then will be better to try out other libs for my use case instead, as the discussion you link fairly warns that, when faking a browser context, other libs expectations might break. And I don't want such potential surprises in the foundation I am setting up for my applications.

**Reply by edemaine** [reply]
That's too bad! We definitely want to make Solid work well in this use case, but it's a little tricky to figure out the best way to support it and SSR at the same time (hence #1104). Hopefully this will be fixed sooner rather than later.

I take it that the elaborate `import` also isn't suitable for your use-case?

**Reply by msteen** [reply]
Ah, good to know. I was afraid I was trying to force Solid in a use case it was never meant for, i.e. a bad fit. But if the aim is to have the Reactivity API be usable standalone as well, i.e. outside HTML rendering, I will definitely give your suggestion a try and see if any potential problems that might arise are actually blocking for me.

**Reply by msteen** [reply]
Did not want to wait until tomorrow to try out the `import` trick, so I just tried, and it works! I do need to work out some issues though, as e.g. TypeScript typing breaks and it requires compilation as ES modules, but those shouldn't be too much of a problem. Thanks for the help!

**Reply by edemaine** [reply]
Great! There's a `solid.cjs` file if you want to use CommonJS instead. I imagine there's a workaround for types... maybe `"types": "solid-js"`? You might come by the Discord and ask about that one (some TS experts in the #typescript channel).

**Reply by msteen** [reply]
I have the TypeScript working too, though there might be better solutions I have yet to try, at least I have a solution that works for now (in a `solid.d.ts` file):

```ts
declare module 'solid-js/dist/solid.cjs' {
    export * from 'solid-js/types';
}
```

**Reply by msteen** [reply]
Unfortunately this fails for subpackages like `solid-js/store`, as it internally imports `solid-js`, which still resolves to the server variant. I tried to circumvent that by creating a local copy of solid-js and modifying the `package.json` file to have the `node` paths point to the `browser` paths, which seems to have worked:

```
console.log(require.resolve('solid-js'));
// [...]\packages\solid-js\dist\solid.cjs
console.log(require.resolve('solid-js/store'));
// [...]\packages\solid-js\store\dist\store.cjs
```

Initially it failed for `solid-js/store` even with the workaround, but that was caused by it (the store) requiring `solid-js` and my customized `package.json` not being in a parent directory of it, so the self-reference resolved to the unmodified `solid-js`.

**Reply by edemaine** [reply]
Ouch, that makes sense. A workaround would be to define import aliases, which might be supported by your bundler. More motivation for #1104... (but to be clear, that is a not-yet-accepted proposal)
